File: inventary-service/src/inventory_client.py
import grpc
import inventory_pb2
import inventory_pb2_grpc

class InventoryClient:
    def __init__(self, host="inventory", port=50051):
        self.channel = grpc.insecure_channel(f"{host}:{port}")
        self.stub = inventory_pb2_grpc.InventoryServiceStub(self.channel)

    def check_stock(self, book_id):
        return self.stub.CheckStock(inventory_pb2.StockRequest(book_id=book_id))

    def reserve_stock(self, book_id, quantity):
        return self.stub.ReserveStock(inventory_pb2.ReserveRequest(book_id=book_id, quantity=quantity))

    def confirm_stock(self, book_id, quantity):
        return self.stub.ConfirmStockReduction(inventory_pb2.ConfirmRequest(book_id=book_id, quantity=quantity))

File: cart-service/grpc/inventoryClient.js
// cart-service/grpc/inventoryClient.js

const grpc = require('@grpc/grpc-js');
const protoLoader = require('@grpc/proto-loader');
const path = require('path');

const PROTO_PATH = path.join(__dirname, 'proto', 'inventory.proto');

const packageDefinition = protoLoader.loadSync(PROTO_PATH, {
  keepCase: true,
  longs: String,
  enums: String,
  defaults: true,
  oneofs: true,
});

const inventoryProto = grpc.loadPackageDefinition(packageDefinition).inventory;

class InventoryClient {
  constructor() {
    // Si existe variable de entorno, la usa. Si no, usa localhost:50051
    const INVENTORY_GRPC_URL = process.env.INVENTORY_GRPC_URL || 'inventory:50051';


    this.client = new inventoryProto.InventoryService(
      INVENTORY_GRPC_URL,
      grpc.credentials.createInsecure()
    );

    console.log(` Inventory gRPC Client conectado a: ${INVENTORY_GRPC_URL}`);
  }

  checkStock(bookId) {
    return new Promise((resolve, reject) => {
      this.client.CheckStock({ book_id: bookId }, (err, resp) => {
        if (err) return reject(err);
        resolve(resp);
      });
    });
  }

  reserveStock(bookId, quantity) {
    return new Promise((resolve, reject) => {
      this.client.ReserveStock({ book_id: bookId, quantity }, (err, resp) => {
        if (err) return reject(err);
        resolve(resp);
      });
    });
  }

  confirmStock(bookId, quantity) {
    return new Promise((resolve, reject) => {
      this.client.ConfirmStockReduction({ book_id: bookId, quantity }, (err, resp) => {
        if (err) return reject(err);
        resolve(resp);
      });
    });
  }

  releaseStock(bookId, quantity) {
    return new Promise((resolve, reject) => {
      this.client.ReleaseStock({ book_id: bookId, quantity }, (err, resp) => {
        if (err) return reject(err);
        resolve(resp);
      });
    });
  }
}

module.exports = new InventoryClient();
